lib/elixir/lib/enum.ex

```diff
@@ -1315,17 +1315,21 @@ defmodule Enum do
   If multiple elements are considered maximal, the first one that was found
   is returned.
 
-  Raises `Enum.EmptyError` if `enumerable` is empty.
+  Calls the provided `empty_fallback` and returns its value if `enumerable`
+  is empty. The default `empty_fallback` raises `Enum.EmptyError`.
 
   ## Examples
 
       iex> Enum.max([1, 2, 3])
       3
 
+      iex> Enum.max([], fn -> 0 end)
+      0
+
   """
-  @spec max(t) :: element | no_return
-  def max(enumerable) do
-    reduce(enumerable, &Kernel.max(&1, &2))
+  @spec max(t, (() -> empty_result)) :: element | empty_result | no_return when empty_result: any
+  def max(enumerable, empty_fallback \\ fn -> raise Enum.EmptyError end) do
```

Comment:
We should have a `when is_function(empty_fallback, 0)` guard here and in `min/2`.
Also it would be better to move the definition of default for `empty_fallback` to bodiless clause.